boot/boot.go

```diff
@@ -0,0 +1,189 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+)
+
+// A BootParticipant is a snap that is involved in a device's boot
+// process.
+type BootParticipant interface {
+	// SetNextBoot will schedule the snap to be used in the next boot. For
+	// base snaps it is up to the caller to select the right bootable base
+	// (from the model assertion).
+	SetNextBoot() error
+	// ChangeRequiresReboot returns whether a reboot is required to switch
+	// to the snap.
+	ChangeRequiresReboot() bool
+}
+
+// A Kernel is a snap.TypeKernel BootParticipant
+type Kernel interface {
+	BootParticipant
+	// RemoveKernelAssets removes the unpacked kernel/initrd for the given
+	// kernel snap.
+	RemoveKernelAssets() error
+	// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
+	// kernel snap, if required, to a versioned bootloader directory so
+	// that the bootloader can use it.
+	ExtractKernelAssets(snap.Container) error
+}
+
+type Model interface {
+	Kernel() string
+	Base() string
+}
+
+// Lookup figures out what the boot participant is for the given arguments, and
+// returns it. The second return value indicates whether no boot participant
+// exists, and if false the first return value will be nil.
+//
+// Currently, on classic, nothing is a boot participant (applicable
+// will always be false).
+func Lookup(s snap.PlaceInfo, t snap.Type, model Model, onClassic bool) (bp BootParticipant, applicable bool) {
+	if onClassic {
+		return nil, false
+	}
+	if t != snap.TypeOS && t != snap.TypeKernel && t != snap.TypeBase {
+		return nil, false
+	}
+
+	if model != nil {
+		switch t {
+		case snap.TypeKernel:
+			if s.InstanceName() != model.Kernel() {
+				return nil, false
+			}
+		case snap.TypeBase, snap.TypeOS:
+			base := model.Base()
+			if base == "" {
+				base = "core"
+			}
+			if s.InstanceName() != base {
+				return nil, false
+			}
+		}
+	}
+
+	cbp := &coreBootParticipant{s: s, t: t}
+	bp = cbp
+	if t == snap.TypeKernel {
+		bp = &coreKernel{cbp}
+	}
+
+	return bp, true
+}
+
+// InUse checks if the given name/revision is used in the
+// boot environment
+func InUse(name string, rev snap.Revision) bool {
+	bootloader, err := bootloader.Find()
+	if err != nil {
+		logger.Noticef("cannot get boot settings: %s", err)
+		return false
+	}
+
+	bootVars, err := bootloader.GetBootVars("snap_kernel", "snap_try_kernel", "snap_core", "snap_try_core")
+	if err != nil {
+		logger.Noticef("cannot get boot vars: %s", err)
+		return false
+	}
+
+	snapFile := filepath.Base(snap.MountFile(name, rev))
+	for _, bootVar := range bootVars {
+		if bootVar == snapFile {
+			return true
+		}
+	}
+
+	return false
+}
+
+var (
+	ErrBootNameAndRevisionAgain = errors.New("boot revision not yet established")
```

Comment:
Perhaps this is existing code but I find the error text and variable name disconnected.